File: ios/MathFriends/MathFriends/Services/ProblemGenerator.swift
//
//  ProblemGenerator.swift
//  MathFriends
//
//  Direct port from TypeScript problem generation logic
//  Generates math problems for all three difficulty levels
//

import Foundation

/// Service for generating math problems
final class ProblemGenerator {
    private let config: AppConfigurationProtocol
    private var rng: RandomNumberGenerator
    private var idGenerator: IDGenerator

    init(
        config: AppConfigurationProtocol = AppConfiguration.shared,
        rng: RandomNumberGenerator = SystemRandomNumberGenerator(),
        idGenerator: IDGenerator = UUIDGenerator()
    ) {
        self.config = config
        self.rng = rng
        self.idGenerator = idGenerator
    }

    /// Generate a problem for the specified difficulty
    func generate(difficulty: Difficulty) -> AnyProblem {
        switch difficulty {
        case .superEasy:
            return AnyProblem(generateSuperEasyProblem())
        case .easy:
            return AnyProblem(generateEasyProblem())
        case .medium:
            return AnyProblem(generateMediumProblem())
        case .hard:
            return AnyProblem(generateHardProblem())
        case .extreme:
            return generateExtremeProblem()
        }
    }

    // MARK: - Super Easy Problems (Simple Addition 0-5)

    private func generateSuperEasyProblem() -> BasicProblem {
        let num1 = rng.nextInt(in: config.superEasyNumberRange)
        let num2 = rng.nextInt(in: config.superEasyNumberRange)
        let answer = Double(num1 + num2)

        return BasicProblem(
            id: idGenerator.generateID(),
            num1: num1,
            num2: num2,
            operation: .addition,
            answer: answer,
            difficulty: .superEasy
        )
    }

    // MARK: - Easy Problems (Addition & Subtraction 0-20)

    private func generateEasyProblem() -> BasicProblem {
        // Randomly choose addition or subtraction
        let operation: Operation = rng.next() > 0.5 ? .addition : .subtraction

        let num1: Int
        let num2: Int
        let answer: Double

        if operation == .addition {
            // Addition: ensure sum ≤ 20
            num1 = rng.nextInt(in: config.easyNumberRange)
            num2 = rng.nextInt(in: 0...(config.easyNumberRange.upperBound - num1))
            answer = Double(num1 + num2)
        } else {
            // Subtraction: ensure no negatives
            num1 = rng.nextInt(in: config.easyNumberRange)
            num2 = rng.nextInt(in: 0...num1)
            answer = Double(num1 - num2)
        }

        return BasicProblem(
            id: idGenerator.generateID(),
            num1: num1,
            num2: num2,
            operation: operation,
            answer: answer,
            difficulty: .easy
        )
    }

    // MARK: - Medium Problems (Multiplication, Division, Powers)

    private func generateMediumProblem() -> BasicProblem {
        let problemTypes: [Operation] = [.multiplication, .division, .power]
        guard let problemType = rng.nextElement(from: problemTypes) else {
            fatalError("Failed to select problem type")
        }

        let num1: Int
        let num2: Int
        let answer: Double

        switch problemType {
        case .multiplication:
            // Multiplication: 0-12 × 0-12
            num1 = rng.nextInt(in: config.mediumMultiplicationRange)
            num2 = rng.nextInt(in: config.mediumMultiplicationRange)
            answer = Double(num1 * num2)

        case .division:
            // Division: Generate quotient first to ensure whole numbers
            num2 = rng.nextInt(in: 1...config.mediumMultiplicationRange.upperBound) // divisor (no division by 0)
            let quotient = rng.nextInt(in: 1...config.mediumMultiplicationRange.upperBound)
            num1 = num2 * quotient // dividend = divisor × quotient
            answer = Double(quotient)

        case .power:
            // Powers: base (2-5), exponent (2-5)
            guard let base = rng.nextElement(from: config.mediumPowerBases) else {
                fatalError("Failed to select power base")
            }
            num1 = base
            num2 = rng.nextInt(in: config.mediumPowerExponentRange)
            answer = pow(Double(num1), Double(num2))

        default:
            fatalError("Unexpected operation type")
        }

        return BasicProblem(
            id: idGenerator.generateID(),
            num1: num1,
            num2: num2,
            operation: problemType,
            answer: answer,
            difficulty: .medium
        )
    }

    // MARK: - Hard Problems (Linear Equations: ax + b = c)

    private func generateHardProblem() -> EquationProblem {
        let x = rng.nextInt(in: config.hardVariableRange)
        let a = rng.nextInt(in: config.hardCoefficientRange)
        let b = rng.nextInt(in: config.hardConstantRange)
        let c = a * x + b

        let bSign = b >= 0 ? "+" : "-"
        let bValue = abs(b)
        let equation = "\(a)x \(bSign) \(bValue) = \(c)"

        return EquationProblem(
            id: idGenerator.generateID(),
            equation: equation,
            answer: Double(x),
            difficulty: .hard
        )
    }

    // MARK: - Extreme Problems (Systems of Equations & Quadratics)

    private func generateExtremeProblem() -> AnyProblem {
        return rng.next() > 0.5
            ? AnyProblem(generateSystemEquationProblem())
            : AnyProblem(generateQuadraticProblem())
    }

    private func generateSystemEquationProblem() -> SystemEquationProblem {
        let x = rng.nextInt(in: config.extremeVariableRange)
        let y = rng.nextInt(in: config.extremeVariableRange)

        var a1 = rng.nextInt(in: 1...6)
        var b1 = rng.nextInt(in: 1...6)
        var a2 = rng.nextInt(in: 1...6)
        var b2 = rng.nextInt(in: 1...6)

        while a1 * b2 - a2 * b1 == 0 {
            b2 = rng.nextInt(in: 1...6)
        }

        if rng.next() > 0.5 { b1 = -b1 }
        if rng.next() > 0.5 { a2 = -a2 }

        let c1 = a1 * x + b1 * y
        let c2 = a2 * x + b2 * y

        let equation1 = "\(formatTerm(a1, "x", true))\(formatTerm(b1, "y", false)) = \(c1)"
        let equation2 = "\(formatTerm(a2, "x", true))\(formatTerm(b2, "y", false)) = \(c2)"

        return SystemEquationProblem(
            id: idGenerator.generateID(),
            equation1: equation1,
            equation2: equation2,
            answerX: Double(x),
            answerY: Double(y),
            difficulty: .extreme
        )
    }

    private func generateQuadraticProblem() -> EquationProblem {
        let r1 = rng.nextInt(in: -8...8)
        let r2 = rng.nextInt(in: -8...8)

        let b = -(r1 + r2)
        let c = r1 * r2
        let answer = min(r1, r2)

        var equation = "x²"
        if b != 0 {
            if b > 0 {
                equation += " + \(b == 1 ? "" : "\(b)")x"
            } else {
                equation += " - \(b == -1 ? "" : "\(abs(b))")x"
            }
        }
        if c != 0 {
            equation += c > 0 ? " + \(c)" : " - \(abs(c))"
        }
        equation += " = 0"

        return EquationProblem(
            id: idGenerator.generateID(),
            equation: equation,
            answer: Double(answer),
            difficulty: .extreme
        )
    }

    // MARK: - Helpers

    private func formatTerm(_ coeff: Int, _ variable: String, _ isFirst: Bool) -> String {
        let absCoeff = abs(coeff)
        let sign = coeff >= 0 ? (isFirst ? "" : " + ") : (isFirst ? "-" : " - ")
        let coeffStr = absCoeff == 1 ? "" : "\(absCoeff)"
        return "\(sign)\(coeffStr)\(variable)"
    }
}

File: client/src/utils/problemGenerator.ts
import { Difficulty } from "../../../shared/src/types/problems";

function generateId(): string {
  return Math.random().toString(36).substring(2, 11);
}

function randomInt(min: number, max: number): number {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function randomChoice<T>(arr: T[]): T {
  return arr[Math.floor(Math.random() * arr.length)];
}

export interface BasicProblem {
  id: string;
  num1: number;
  num2: number;
  operation: "+" | "-" | "*" | "/" | "^";
  answer: number;
  difficulty: Difficulty;
  display: string;
}

export interface EquationProblem {
  id: string;
  equation: string;
  answer: number;
  difficulty: "hard" | "extreme";
}

export interface ComparisonProblem {
  id: string;
  expression: string;
  expressionResult: number;
  comparisonNumber: number;
  answer: number; // 0 = less than, 1 = greater than, 2 = equal to
  difficulty: "comparison";
  display: string;
}

export interface SystemEquationProblem {
  id: string;
  equation1: string;
  equation2: string;
  answerX: number;
  answerY: number;
  difficulty: "extreme";
}

export type Problem =
  | BasicProblem
  | ComparisonProblem
  | EquationProblem
  | SystemEquationProblem;

function generateSuperEasyProblem(): BasicProblem {
  const num1 = randomInt(0, 5);
  const num2 = randomInt(0, 5);
  const answer = num1 + num2;

  return {
    id: generateId(),
    num1,
    num2,
    operation: "+",
    answer,
    difficulty: "superEasy",
    display: `${num1} + ${num2} = ?`,
  };
}

function generateEasyProblem(): BasicProblem {
  const operation = Math.random() > 0.5 ? "+" : "-";

  let num1: number, num2: number, answer: number;

  if (operation === "+") {
    num1 = randomInt(0, 20);
    num2 = randomInt(0, 20 - num1);
    answer = num1 + num2;
  } else {
    num1 = randomInt(0, 20);
    num2 = randomInt(0, num1);
    answer = num1 - num2;
  }

  return {
    id: generateId(),
    num1,
    num2,
    operation: operation as "+" | "-",
    answer,
    difficulty: "easy",
    display: `${num1} ${operation} ${num2} = ?`,
  };
}

function generateMediumProblem(): BasicProblem {
  const problemTypes = ["*", "/", "^"];
  const problemType = randomChoice(problemTypes);

  let num1: number, num2: number, answer: number;

  if (problemType === "*") {
    num1 = randomInt(0, 12);
    num2 = randomInt(0, 12);
    answer = num1 * num2;
  } else if (problemType === "/") {
    num2 = randomInt(1, 12);
    const quotient = randomInt(1, 12);
    num1 = num2 * quotient;
    answer = quotient;
  } else {
    num1 = randomChoice([2, 3, 4, 5]);
    num2 = randomInt(2, 5);
    answer = Math.pow(num1, num2);
  }

  const display =
    problemType === "^"
      ? `${num1}^${num2} = ?`
      : `${num1} ${problemType} ${num2} = ?`;

  return {
    id: generateId(),
    num1,
    num2,
    operation: problemType as "*" | "/" | "^",
    answer,
    difficulty: "medium",
    display,
  };
}

function generateHardProblem(): EquationProblem {
  const x = randomInt(-10, 10);
  const a = randomInt(1, 10);
  const b = randomInt(-20, 20);
  const c = a * x + b;

  const equation = `${a}x ${b >= 0 ? "+" : "-"} ${Math.abs(b)} = ${c}`;

  return {
    id: generateId(),
    equation,
    answer: x,
    difficulty: "hard",
  };
}

function formatTerm(coeff: number, variable: string, isFirst: boolean): string {
  const absCoeff = Math.abs(coeff);
  const sign = coeff >= 0 ? (isFirst ? "" : " + ") : isFirst ? "-" : " - ";
  const coeffStr = absCoeff === 1 ? "" : `${absCoeff}`;
  return `${sign}${coeffStr}${variable}`;
}

function generateSystemEquationProblem(): SystemEquationProblem {
  const x = randomInt(-10, 10);
  const y = randomInt(-10, 10);

  let a1 = randomInt(1, 6);
  let b1 = randomInt(1, 6);
  let a2 = randomInt(1, 6);
  let b2 = randomInt(1, 6);

  while (a1 * b2 - a2 * b1 === 0) {
    b2 = randomInt(1, 6);
  }

  if (Math.random() > 0.5) b1 = -b1;
  if (Math.random() > 0.5) a2 = -a2;

  const c1 = a1 * x + b1 * y;
  const c2 = a2 * x + b2 * y;

  const equation1 = `${formatTerm(a1, "x", true)}${formatTerm(b1, "y", false)} = ${c1}`;
  const equation2 = `${formatTerm(a2, "x", true)}${formatTerm(b2, "y", false)} = ${c2}`;

  return {
    id: generateId(),
    equation1,
    equation2,
    answerX: x,
    answerY: y,
    difficulty: "extreme",
  };
}

function generateQuadraticProblem(): EquationProblem {
  // Generate quadratic x² + bx + c = 0 with integer roots r1, r2
  const r1 = randomInt(-8, 8);
  const r2 = randomInt(-8, 8);

  // x² - (r1+r2)x + r1*r2 = 0
  const b = -(r1 + r2);
  const c = r1 * r2;

  // Use the smaller root as the answer
  const answer = Math.min(r1, r2);

  let equation = "x²";
  if (b !== 0) {
    equation +=
      b > 0 ? ` + ${b === 1 ? "" : b}x` : ` - ${b === -1 ? "" : Math.abs(b)}x`;
  }
  if (c !== 0) {
    equation += c > 0 ? ` + ${c}` : ` - ${Math.abs(c)}`;
  }
  equation += " = 0";

  return {
    id: generateId(),
    equation,
    answer,
    difficulty: "extreme",
  };
}

function generateComparisonProblem(): ComparisonProblem {
  const operation = randomChoice(["+", "-", "*"] as const);

  let num1: number, num2: number, expressionResult: number;

  if (operation === "+") {
    num1 = randomInt(1, 20);
    num2 = randomInt(1, 20);
    expressionResult = num1 + num2;
  } else if (operation === "-") {
    num1 = randomInt(5, 30);
    num2 = randomInt(1, num1 - 1);
    expressionResult = num1 - num2;
  } else {
    num1 = randomInt(2, 10);
    num2 = randomInt(2, 10);
    expressionResult = num1 * num2;
  }

  const operationSymbol = operation === "*" ? "\u00D7" : operation;
  const expression = `${num1} ${operationSymbol} ${num2}`;

  // Generate a comparison number: greater, less, or equal to the result
  let comparisonNumber: number;
  const roll = Math.random();
  if (roll < 0.33) {
    // Equal
    comparisonNumber = expressionResult;
  } else if (roll < 0.66) {
    // Greater
    comparisonNumber = expressionResult + randomInt(1, 10);
  } else {
    // Less
    comparisonNumber = expressionResult - randomInt(1, 10);
  }

  // 0 = less than, 1 = greater than, 2 = equal to
  let answer: number;
  if (comparisonNumber > expressionResult) {
    answer = 1;
  } else if (comparisonNumber < expressionResult) {
    answer = 0;
  } else {
    answer = 2;
  }

  return {
    id: generateId(),
    expression,
    expressionResult,
    comparisonNumber,
    answer,
    difficulty: "comparison",
    display: `Is ${comparisonNumber} greater than, less than, or equal to ${expression} ?`,
  };
}

function generateExtremeProblem(): Problem {
  return Math.random() > 0.5
    ? generateSystemEquationProblem()
    : generateQuadraticProblem();
}

export function generateProblem(difficulty: Difficulty): Problem {
  switch (difficulty) {
    case "superEasy":
      return generateSuperEasyProblem();
    case "easy":
      return generateEasyProblem();
    case "comparison":
      return generateComparisonProblem();
    case "medium":
      return generateMediumProblem();
    case "hard":
      return generateHardProblem();
    case "extreme":
      return generateExtremeProblem();
    default:
      throw new Error(`Unknown difficulty: ${difficulty}`);
  }
}
